src/yamerl_node_map.erl

```diff
@@ -142,3 +133,72 @@ node_pres(Node) ->
 
 node_as_proplist_or_map(#yamerl_constr{ext_options = Options}) ->
     proplists:get_value(map_node_format, Options, proplist).
+
+new_builder(Constr) ->
+    Format = node_as_proplist_or_map(Constr),
+    Map = #map_builder{
+        format = Format,
+        state = none,
+        keys = []},
+    case Format of
+        proplist ->
+            Map#map_builder{data = []};
+        map ->
+            Map#map_builder{data = maps:new()}
+    end.
+
+finalize_builder(#map_builder{data = Map}) when is_list(Map) ->
+    lists:reverse(Map);
+finalize_builder(#map_builder{data = Map}) ->
+    Map.
+
+set_kv(#yamerl_constr{detailed_constr = false},
+       Key,
+       Value,
+       #map_builder{format = proplist, data = Map} = Builder) ->
+    case lists:keymember(Key, 1, Map) of
+        true ->
+            Builder#map_builder{state = none, data = lists:keyreplace(Key, 1, Map, {Key, Value})};
+        false ->
+            Builder#map_builder{state = none, data = [{Key, Value} | Map]}
+    end;
+set_kv(#yamerl_constr{detailed_constr = false},
+       Key,
+       Value,
+       #map_builder{format = map, data = Map} = Builder) ->
+    Builder#map_builder{state = none, data = maps:put(Key, Value, Map)};
+
+set_kv(#yamerl_constr{detailed_constr = true},
+       Key,
+       Value,
+       #map_builder{format = proplist, keys = Keys, data = Map} = Builder) ->
+    case find(Key, Keys) of
+        none ->
+            RawKey = erlang:delete_element(#yamerl_str.pres, Key),
```

Comment:
Could you please isolate the conversion from `Key` to `RawKey` in a dedicated function? I believe it would be clearer what the intent is.